Add isRootDropTarget parameter to renderEmptyState in GridList component

gennadiipel opened the pull request:
Add `isRootDropTarget` parameter to `renderEmptyState` in `GridList` component

Closes #5210

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [x] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

<!--- Include instructions to test this pull request -->

## 🧢 Your Project:

<!--- Company/project for pull request -->

reidbarber commented on packages/react-aria-components/src/GridList.tsx:
```diff
@@ -54,7 +54,7 @@ export interface GridListProps<T> extends Omit<AriaGridListProps<T>, 'children'>
   dragAndDropHooks?: DragAndDropHooks,
   /** Provides content to display when there are no items in the list. */
-  renderEmptyState?: () => ReactNode
+  renderEmptyState?: (isRootDropTarget: boolean) => ReactNode
```
```suggestion
  renderEmptyState?: (props: {isDropTarget: boolean, isFocused: boolean, isFocusVisible: boolean, state: ListState<unknown>}) => ReactNode
```

reidbarber reviewed (changes requested):
I like this idea a lot!

I say we go ahead and pass all the render props in (except for `isEmpty` since it is implied). And we can use one argument with everything.

reidbarber commented:
GET_BUILD

snowystinger commented on packages/react-aria-components/src/GridList.tsx:
```diff
@@ -176,7 +176,12 @@ function GridListInner<T extends object>({props, collection, gridListRef: ref}:
     // a bug that makes grid elements with display: contents hidden to screen readers.
     // https://bugs.webkit.org/show_bug.cgi?id=239479
-    let content = props.renderEmptyState();
+    let content = props.renderEmptyState({
```
seems like a fine approach to me
we will need this in ListBox, Table, and TagGroup as well in order to be consistent
also, some form of a test/storybook example would be appreciated if you do not mind https://react-spectrum.adobe.com/contribute.html#pull-requests

gennadiipel commented:
@snowystinger @reidbarber Hello, I updated `Table`, `ListBox` and `TagGroup`. Also updated docs a little and added example of reorderable tables including PR's changes to storybook. Please take a look.

snowystinger commented:
GET_BUILD

snowystinger commented on packages/react-aria-components/docs/ListBox.mdx:
```diff
@@ -850,7 +850,7 @@ Use the `renderEmptyState` prop to customize what the `ListBox` will display if
   aria-label="Search results"
   /*- begin highlight -*/
-  renderEmptyState={() => 'No results found.'}
+  renderEmptyState={({isDropTarget, state, isFocused, isFocusVisible}) => <span style={{color: isDropTarget ? 'red' : 'black'}}>No results found.</span>}
```
these are great, though in dark mode, it's a little hard to see
<img width="714" alt="Screenshot 2023-10-10 at 1 59 16 PM" src="https://github.com/adobe/react-spectrum/assets/698229/c7555f2a-dcfd-471f-8afa-a9892eb8e8fd">

Instead, we should use one of the variables available to us in our docs, `--spectrum-alias-text-color`
There is another issue, though: I cannot drop anything on this example, so it never goes to red.

As a result, I think it should be in the example a little further down https://github.com/adobe/react-spectrum/blob/main/packages/react-aria-components/docs/ListBox.mdx#L1101
And instead of changing colors (esp to red which is negative), let's swap the string. Something like
```
isDropTarget ? 'This will not be uploaded' : 'Drop items here'
```

gennadiipel replied:
Yep, updated docs like that in all of the cases and also added for `GridList` (forgot this one)

reidbarber commented:
GET_BUILD

reidbarber reviewed (dismissed):
LGTM, thanks!!
